lambda/matching.ml

```diff
@@ -3431,7 +3439,52 @@ and combine_handlers ~scopes repr partial ctx (v, str, arg) first_match rem =
          repr)
       partial ctx first_match rem
   in
-  (bind_check str v arg lam, total)
+  let jumps =
+    (* If the Lambda expression [arg] to access the first argument is
+       a mutable field read, then its binding and evaluation may be
+       emitted in different calls to [combine_handlers] on the same
+       column. Consider for example:
+
+         type ('a, 'b) mut_second = { immut : 'a; mutable mut : 'b; }
+
+         function
+         | {immut = false; mut = None} -> -1
+         | {immut = true ; mut = None} -> 0
+         | {immut = _ ;    mut = Some n} -> n
+
+       When compiling this example, [immut] will be matched first, and
+       each case will perform a [None] check and also jump to a shared
+       exit handler containing the [Some n] clause. The field access
+       to the [mut] field will be emitted three times, in each branch
+       of the switch and in the shared handler.
+
+       In the general case, the value of the mutable field may change
+       between the reads (due to a [when] guard or even a race from
+       another thread or domain), so we must be careful not to
+       propagate context information that could have become
+       incorrect. We "fix" the context information on mutable arguments
+       by calling [Context.erase_first_col] below.
+    *)
+    let mut =
+      (* This is somewhat of a hack: we notice that a pattern-matching
+         argument is mutable (its value can change if evaluated
+         several times) exactly when it is bound as StrictOpt. Alias
+         bindings are obviously pure, but Strict bindings are also only
+         used in the pattern-matching compiler for expressions that give
+         the same value when evaluated twice.
+         An alternative would be to track 'mutability of the field'
+         directly.
```

Comment:
I have, in fact, implemented such a change in https://github.com/gasche/ocaml/commit/53e71e95ac4c6d97b690082c1c6edfe7f6f27432, which is part of the `matching-bug-mutable-positions` branch that is planned as one of the follow-ups to the present PR as described in the overview comment https://github.com/ocaml/ocaml/issues/7241#issuecomment-1722226025 .

In the overview comment I explain that not all cases of unsoundness are fixed by the present PR, some other are due to incorrect totality information, and tracking mutability of arguments is important to fix those extra cases. I don't have the information in brain memory and haven't looked at your example in details yet, so at this point I am not sure whether your example is exactly a case that should be fixed by the "second half" of the bug (incorrect totality information), or if it is a different instance of the "first half" of the bug (incorrect context information) that needs the present PR to be fixed.

Reply:
At least I can confirm that my branch [matching-bug-fix-totality](https://github.com/gasche/ocaml/commits/matching-bug-fix-totality), which is supposed to contain both "halves" of the fix (and to need a rebase), also fixes the segfault on your example.

Reply:
My feeling when I wrote this comment was that it is a different instance of the "first half" of the bug, and is unrelated to incorrect totality information. The https://github.com/gasche/ocaml/commit/53e71e95ac4c6d97b690082c1c6edfe7f6f27432 commit does not appear to be present as-is in `matching-bug-fix-totality` yet (you say elsewhere that branch needs a rebase) so I can't confirm yet.

Would you object to adding this as a test case in the `matching-bug-fix-totality` branch so we can be sure to remember this?

Reply:
> Would you object to adding this as a test case in the matching-bug-fix-totality branch so we can be sure to remember this?

Not at all. In case you are interested, my working repository uses the following test, which is relatively close in spirit but may be oversimplified and miss some aspects of your test:

```ocaml
(* This example exhibits unsoundness due to the totality issue *)

let r = ref (Some 3)

let _ =
  match Some r with
  | Some { contents = None } -> 0
  | _ when (r := None; false) -> 1
  | Some { contents = Some n } -> n
  | None -> 3
```

Reply:
That test case looks good, in that it seems close in spirit to my original test case and it segfaults in the present PR.

Reply:
I had completely forgotten about it yesterday, but #12548 that we already merged contains this testcase (with its current known-broken behavior recorded in the testsuite) and others.

Reply:
Thanks. I find your explanation convincing: the segfaulting program in my original comment indeed seems to be "about totality" rather than about wrong contexts. However, I still have a vague concern about the fact that `get_expr_args_record`, etc., `_`-ignores the mutability of its argument. It will be easier for me to think about this question in [matching-bug-fix-totality](https://github.com/gasche/ocaml/commits/matching-bug-fix-totality) post-rebase, so I'll shelve this thought until then.

Reply:
In my perspective, "arguments" in the pattern-matching machinery are either variables (the typical case) or non-variable expressions that are produced by the `get_expr_args*` functions by projecting out (in some way) another argument that is itself a variable. So for example, `match x with { f = { g = ... } } -> ...` will have arguments `x`, then `x.f` that will in turn be bound to `y`, then `y.g` that will be bound to some `z`.

`x.f.g` as a complex expression is never used as an argument in the generated code -- but this is not so easy to check by reading the code, it comes from the fact `compile_match_nonempty` replaces non-variable arguments by variable arguments before calling `combine_handlers`, which in turn will eventually create the "division" out of the now-variable argument list in `do_compile_matching`.

So when I write in the comment that an argument is mutable (maybe "impure" would be the better term), defined as "its value can change if evaluated several time", well `y.g` can change if evaluated several times if `g` is a mutable field, but that property does *not* depend on whether `f` itself was a mutable field. So I do not expect that we need to transitively propagate mutability for this part of the bug, caused by sharing context information across several different evaluations of a same argument expression. (On the other hand, the transitive form of mutability that you have in mind shows up in the totality part, as a criterion for when to distrust the type-checker approximation of totality: we degrade Total into Partial when under a mutable context (transitively).)

Reply:
I don't know how you want to proceed: we could try to prepare all changes at once for merging, so that you can have a detailed complete picture, or we could try to review the changes step by step, and try to merge a PR before moving to followup changes. The latter is less work for me (because I expect reviews to suggest further changes that may require non-trivial rebases of followup changes), but it makes it harder to have a global view. On the other hand personally I think that reviewing bits by bits is easier and more comfortable -- there are less chances of letting things slip by just because it is too large to be looked at properly.

(In either scenario, the precise process I would have in mind is to have you "Approve" of the PR once you are convinced that it is correct and you find the implementation reasonable/maintenable/nice, and then we would look for a maintainer willing to have a second look or to approve on your behalf.)

Reply:
I'm in favor of reviewing in pieces as well. I intend on clicking "Approve" soon &mdash; I'm just verifying one or two more things.

Reply:
And your explanation here covers what I was missing before, I think:

> x.f.g as a complex expression is never used as an argument in the generated code -- but this is not so easy to check by reading the code, it comes from the fact compile_match_nonempty replaces non-variable arguments by variable arguments before calling combine_handlers, which in turn will eventually create the "division" out of the now-variable argument list in do_compile_matching.